File: runme/mcgen/local/datan/datsrc/sqchi2.c
#include "datsrc.h"

double LIBFUNC sqchi2_(double const P_T *p, integer const P_T *n)
{
    /* Table of constant values */

    integer c__0 = 0;

    /* System generated locals */
    double ret_val;

    /* Local variables */
    double xzero, x0, x1, epsiln;

/* boundary of range */
    if (*p >= 1.) {
        ret_val = 1e10;
    }
    if (*p <= 0.) {
        ret_val = 0.;
    }
/* normal range */
    if (*p < 1. && *p > 0.) {
        x1 = (double) (*n);
        x0 = .5 * x1;
        auxzbr_(&x0, &x1, szchi2_, p, n, &c__0);
        auxzfn_(&x0, &x1, &xzero, szchi2_, p, n, &c__0, &epsiln);
        ret_val = xzero;
    }
    return ret_val;
} /* sqchi2_ */

/* ----------------------------------------------------------------- */
double LIBFUNC szchi2_(double const P_T *x, double const P_T *p,
                      integer const P_T *n, integer const P_T *ndum)
{
    /* System generated locals */
    double ret_val;

    /* Local variables */

/* returns P minus cumulative chisquared distribution of (X,N) */
    ret_val = *p - scchi2_(x, n);
    return ret_val;
} /* szchi2_ */

File: runme/mcgen/local/datan/datsrc/sqchi2.f
      DOUBLE PRECISION FUNCTION SQCHI2(P,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER (BIG=1D10,ONE=1.D0, ZERO=0.D0)
      EXTERNAL SZCHI2
C boundary of range
      IF(P.GE.ONE) SQCHI2=BIG
      IF(P.LE.ZERO) SQCHI2=ZERO
C normal range
      IF(P.LT.ONE .AND. P.GT.ZERO) THEN
        X1=DBLE(N)
        X0=DBLE(0.5)*X1
        CALL AUXZBR(X0,X1,SZCHI2,P,N,0)
        CALL AUXZFN(X0,X1,XZERO,SZCHI2,P,N,0,EPSILN)
        SQCHI2=XZERO
      END IF
      END
C-----------------------------------------------------------------
      DOUBLE PRECISION FUNCTION SZCHI2(X,P,N,NDUM)
C returns P minus cumulative chisquared distribution of (X,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      SZCHI2=P-SCCHI2(X,N)
      END
